### PR TITLE
Initialize-response tests

### LINKED ISSUE #105: Test initialization response
It would appear that `ghcide` is not advertising its code action capabilities (emacs-lsp/lsp-ui/issues/319).

Consequently, I would like to add some tests which verify exactly what `ghcide` has to say at the time it is initialized. Could some kind soul who is more familiar with the APIs of `Haskell.Language.LSP` and `Development.IDE.LSP` (and whatever else might be useful) give me some pointers on how best to structure these?

### PR DESCRIPTION
**Pull request by jacg**
These tests document and monitor the evolution of the capabilities
announced by the server in the initialize response.

Currently the server advertises almost no capabilities. Out of 23
top-level categories, the only 3 which are announced are

+ text document sync
+ hover
+ goto definition

At the very least code actions are known to be provided, but are not
announced in the initialize response.

Closes #105

### CODE REVIEW

File: test/exe/Main.hs [hunk 1]
```diff
@@ -30,11 +30,74 @@ main = defaultMain $ testGroup "HIE"
...
+    , chk "NO completion"               _completionProvider  Nothing
+    , chk "NO signature help"        _signatureHelpProvider  Nothing
+    , chk "   goto definition"          _definitionProvider (Just True)
+    , chk "NO goto type definition" _typeDefinitionProvider  Nothing
+    , chk "NO goto implementation"  _implementationProvider  Nothing
+    , chk "NO find references"          _referencesProvider  Nothing
+    , chk "NO doc highlight"     _documentHighlightProvider  Nothing
+    , chk "NO doc symbol"           _documentSymbolProvider  Nothing
+    , chk "NO workspace symbol"    _workspaceSymbolProvider  Nothing
+    , chk "NO code action"              _codeActionProvider  Nothing -- available but not declared !
```
[the remark this answers is not in the document]
Comment (jacg) [reply test/exe/Main.hs]: Could you give some more hints? I had a look around there, but couldn't figure out where exactly to hook in.

[the remark this answers is not in the document]
Comment (cocreature) [reply test/exe/Main.hs]: Of course, happy to try to help! `runLanguageServer` takes the LSP options as the first argument, currently set to `def`, see https://hackage.haskell.org/package/haskell-lsp-0.16.0.0/docs/Language-Haskell-LSP-Core.html#t:Options. That type has a `codeActionProvider` field that allows us to set the `CodeActionOptions` https://hackage.haskell.org/package/haskell-lsp-types-0.16.0.0/docs/Language-Haskell-LSP-Types.html#t:CodeActionOptions.

[the remark this answers is not in the document]
Comment (jacg) [reply test/exe/Main.hs]: Thanks. This leads to some more questions:

+ Should we just add the `CodeActionOptions` to that argument, right there, or is there a better place to do it. For example, it seems that the text document sync options come from  https://github.com/digital-asset/ghcide/blob/master/src/Development/IDE/LSP/LanguageServer.hs#L185 and I haven't found where hover and goto-definiton (the other two declared capabilities) are set. 

+ Aside: I'm confused by the intersection of the 13 fields in https://hackage.haskell.org/package/haskell-lsp-0.16.0.0/docs/Language-Haskell-LSP-Core.html#t:Options with the 23 in https://hackage.haskell.org/package/haskell-lsp-types-0.16.0.0/docs/Language-Haskell-LSP-Types.html#t:InitializeResponseCapabilitiesInner

+ There is a whole bunch of `CodeActionKind`s to choose from. How do I know which ones apply?

+ Aside: why does `_codeActionKinds :: Maybe [CodeActionKind]`? Why the `Maybe` if the list can simply be empty? Is there some semantic difference here between `Nothing` and `Just []`?

[the remark this answers is not in the document]
Comment (cocreature) [reply test/exe/Main.hs]: * Given that there is no way to turn off code actions atm, `modifyOptions` seems like a better place given that users then don’t accidentally pass in the wrong options to `runLanguageServer`.
* `haskell-lsp` deduces a lot of the capabilities from which handlers you set to a `Just` value, see https://github.com/alanz/haskell-lsp/blob/7861b970c5da2ed0f50b9b35533610a43571b036/src/Language/Haskell/LSP/Core.hs#L841
* I believe we only emit `QuickFixes` at the moment, see https://github.com/digital-asset/ghcide/blob/986bc04014ecf144f4de97c206be2724ed11f9f0/src/Development/IDE/LSP/CodeAction.hs#L40
* I don’t think there is a semantic difference. The types in `haskell-lsp` follow the LSP spec very closely and the LSP spec allows for this field to not be present.

Note that the [LSP spec](https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/) also defines that the server must only announce the capability if the client requested it:
```
The `CodeActionOptions` return type is only
valid if the client signals code action literal support via the property
`textDocument.codeAction.codeActionLiteralSupport`
```
so we need to check the capabilities and if it’s not in there probably just set `CodeActionOptionsStatic True`. We do something similar for the progress option which you can probably use as a reference.

[the remark this answers is not in the document]
Comment (jacg) [reply test/exe/Main.hs]: What I take away from this is that `modifyOptions` should set `codeActionProvider` to either `CodeActionOptions (Just [CodeActionQuickFix])` or `CodeActionOptionStatic True` depending on `isJust $ _codeActionLiteralSupport =<< _codeAction =<< _textDocument clientCapabilities`.

Is that right?

If so, then what I don't see is how to make those client capabilities available to `modifyOptions` without digging a huge trench through `runLanguageServer`. Can you suggest the least disruptive way of organizing the plumbing?

[the remark this answers is not in the document]
Comment (cocreature) [reply test/exe/Main.hs]: That’s a very good point, I’m not actually sure how you are supposed to handle this given the fact that `haskell-lsp` doesn’t allow you to set options from `onStartup`. I guess, for now we could probably just set `CodeActionOptionStatic True` which if I understand the spec correctly is always fine even if the client doesn’t specify it in its capabilities.

**Pull request merged**